contracts/Market.sol

```diff
@@ -1929,6 +1771,168 @@ contract Market is
+   * @notice Allocates the removals, amounts, and suppliers needed to fulfill the purchase drawing only from
+   * the specific supplier and vintages specified.
+   * @param amount The number of carbon removals to purchase.
+   * @param supplier The supplier from which to purchase carbon removals.
+   * @param vintages The vintages from which to purchase carbon removals.
+   * @return countOfRemovalsAllocated The number of distinct removal IDs used to fulfill this order.
+   * @return ids An array of the removal IDs being drawn from to fulfill this order.
+   * @return amounts An array of amounts being allocated from each corresponding removal token.
+   */
+  function _allocateSupplySingleSupplierSpecificVintages(
+    uint256 amount,
+    address supplier,
+    uint256[] memory vintages
+  )
+    private
+    returns (
+      uint256 countOfRemovalsAllocated,
+      uint256[] memory ids,
+      uint256[] memory amounts
+    )
+  {
+    RemovalsByYear storage supplierRemovalQueue = _listedSupply[supplier];
+    uint256 countOfSuppliersListedRemovals = 0;
+    for (
+      uint256 vintage = supplierRemovalQueue.earliestYear;
+      vintage <= supplierRemovalQueue.latestYear;
+      ++vintage
+    ) {
+      countOfSuppliersListedRemovals += supplierRemovalQueue
+        .yearToRemovals[vintage]
+        .length();
+    }
+    if (countOfSuppliersListedRemovals == 0) {
+      revert InsufficientSupply();
+    }
+    ids = new uint256[](countOfSuppliersListedRemovals);
+    amounts = new uint256[](countOfSuppliersListedRemovals);
+    uint256 remainingAmountToFill = amount;
+
+    for (uint256 i = 0; i < countOfSuppliersListedRemovals; ++i) {
+      uint256 removalId = supplierRemovalQueue
+        .getNextRemovalForSaleFromVintages({vintages: vintages});
+      // if removalId is 0 here, this supplier doesn't have any more removals from the valid vintages requested,
+      // so we can end the loop
+      if (removalId == 0) {
+        break;
+      }
+      uint256 removalAmount = _removal.balanceOf({
+        account: address(this),
+        id: removalId
+      });
+      if (remainingAmountToFill < removalAmount) {
+        /**
+         * The order is complete, not fully using up this removal, don't check about removing active supplier.
+         */
+        ids[countOfRemovalsAllocated] = removalId;
+        amounts[countOfRemovalsAllocated] = remainingAmountToFill;
+        remainingAmountToFill = 0;
+      } else {
+        /**
+         * We will use up this removal while completing the order, move on to next one.
+         */
+        ids[countOfRemovalsAllocated] = removalId;
+        amounts[countOfRemovalsAllocated] = removalAmount; // this removal is getting used up
+        remainingAmountToFill -= removalAmount;
+        supplierRemovalQueue.remove({removalId: removalId});
+        /**
+         * If the supplier is out of supply, remove them from the active suppliers.
+         */
+        if (supplierRemovalQueue.isEmpty()) {
```

Comment:
Probably worth making sure this missing coverage is hit

Reply:
interesting... I think this and the missing coverage above for `getNextRemovalForSaleFromVintages` both have to do with the fact that these are library functions that are part of the diff in this PR and thus getting picked up by codecov (whereas the many other library calls we use are not).  I believe that because I know there are tests that hit both of these lines for sure.  I wonder what the difference is between calling these functions from a test harness v.s. as part of actual contract code in regular tests? do you have any sense for that?  it's annoying

Reply:
Yeah. I don't remember the exact reason why one is picked up and the other not, but there are several foundry coverage GitHub issues explaining it

Reply:
- https://github.com/foundry-rs/foundry/pull/3128
- https://github.com/foundry-rs/foundry/issues/2567
Both of the above contain several related linked issues describing the issue and workarounds. It's how I learned to fix arraylib coverage. 

Not suggesting we add new tests and fix this issue here (assuming this is not net new code that is also uncovered by a correct test case using a harness), just documenting it.